packages/Chem/src/chem-searches.ts

```diff
@@ -84,116 +90,104 @@ function _chemGetDiversities(limit: number, molStringsColumn: DG.Column, fingerp
-  const colVerTag = '.' + fingerprintsType + '.Version';
 
-  if (col.getTag(colNameTag) &&
-      col.getTag(colVerTag) == String(col.version) &&
-      col.dataFrame) {
-    const fingCol = col.dataFrame.columns.byName(col.getTag(colNameTag));
-    if (fingCol)
-      return fingCol.toList();
+function checkForSavedColumn(col: DG.Column, colTag: Fingerprint | string): DG.Column<any> | null{
+  const savedColName = '~' + col.name + '.' + colTag;
+  const colVerTag = '.' + colTag + '.Version';
+
+  if (col.dataFrame && col.dataFrame?.col(savedColName) &&
+      col.getTag(colVerTag) == String(col.version)) {
+    const savedCol = col.dataFrame.columns.byName(savedColName);
+    return savedCol;
   }
   return null;
 }
 
-function saveFingerprintsToCol(col: DG.Column, fgs: Uint8Array[],
-  fingerprintsType: Fingerprint, createdMols: boolean): void {
+function saveColumns(col: DG.Column, data: ((Uint8Array | null)[] | (string | null)[])[],
+  tags: string[], colTypes: DG.COLUMN_TYPE[]): void {
+  
+  for (let i = 0; i < data.length; i++)
+    saveColumn(col, data[i], tags[i], colTypes[i]);
+
+  const colVersion = col.version + data.length;
+  for (let i = 0; i < data.length; i++)
+    col.setTag('.' + tags[i] + '.Version', String(colVersion));
+}
+
+
+function saveColumn(col: DG.Column, data: (Uint8Array | null)[] | (string | null)[],
+  tagName: string, colType: DG.COLUMN_TYPE): void {
   if (!col.dataFrame)
     throw new Error('Column has no parent dataframe');
 
-  const colNameTag = '.' + fingerprintsType + '.Column';
-  const colVerTag = '.' + fingerprintsType + '.Version';
+  const savedColumnName = '~' + col.name + '.' + tagName;
+  const newCol: DG.Column<Uint8Array> = col.dataFrame.columns.getOrCreate(savedColumnName, colType, data.length);
 
-  const fingerprintColumnName = col.getTag(colNameTag) ??
-    '~' + col.name + fingerprintsType + 'Fingerprints';
-  const df = col.dataFrame;
-  const newCol: DG.Column<Uint8Array> = df.columns.getOrCreate(
-    fingerprintColumnName, DG.COLUMN_TYPE.BYTE_ARRAY, fgs.length);
-
-  newCol.init((i) => fgs[i]);
+  newCol.init((i) => data[i]);
+}
 
-  col.setTag(colNameTag, fingerprintColumnName);
-  if (createdMols)
-    col.setTag(FING_COL_TAGS.molsCreatedForVersion, String(col.version + 3));
-  col.setTag(colVerTag, String(col.version + 2));
-  col.setTag(FING_COL_TAGS.invalidatedForVersion, String(col.version + 1));
+async function invalidateAndSaveColumns(molCol: DG.Column, fingerprintsType: Fingerprint,
+  returnSmiles?: boolean): Promise<IFpResult> {
+  const fpRes = await (await getRdKitService()).getFingerprints(fingerprintsType,  molCol.toList(), returnSmiles);
+  returnSmiles ?
+    saveColumns(molCol, [fpRes.fps, fpRes.smiles!], [fingerprintsType, canonicalSmilesColName],
+      [DG.COLUMN_TYPE.BYTE_ARRAY, DG.COLUMN_TYPE.STRING]):
+    saveColumns(molCol, [fpRes.fps], [fingerprintsType], [DG.COLUMN_TYPE.BYTE_ARRAY]);
+  chemEndCriticalSection();
```

Comment:
Typically, you want to endXXX in the same code block (same function) where you beginXXX. Any reason to keep it the way it is?